### ISSUE TITLE
Research and recommendations per Plexity

### ISSUE THREAD

**Issue by noelsaw1** [HiQS-Labs/XYZ-forge#612]
## Overview & Core Premise

Adding Claude Code or Codex as a "guess-the-next-step" source turns the design into a **teacher–student decision system**: the frontier model is a teacher/proposer, while the 28B Cactus Needle model becomes a student, personalized policy, or re-ranker.

The key is to treat the frontier model’s next-step guess as **additional evidence and candidate generation, not as unquestioned ground truth**. The operator's own eventual action and downstream outcome should remain the highest-value labels.

---

## Method Names & Terminology

**Precise Description:**
> *Human-validated teacher–student distillation for personalized SDLC next-action prediction*

| Emphasis | Useful Term |
|---|---|
| Frontier model provides targets/examples for Cactus Needle | Knowledge distillation |
| Frontier model supplies candidate next moves | LLM-assisted candidate generation |
| Your selection ranks teacher proposals | Human-in-the-loop preference learning |
| Cactus Needle learns your actual next action | Behavioral cloning / imitation learning |
| System chooses among actions based on state and outcome | Contextual bandit |
| Multiple models provide suggestions | Ensembled proposal generation or multi-teacher setup |

Classical knowledge distillation transfers behavior from a more capable teacher into a less costly student; here that can be direct output imitation, synthetic-data generation, or comparative preference/ranking signals.

---

## Recommended Architecture

Use a **generate → normalize → rank → choose → log → learn** loop:

```text
Current SDLC state
  ├── Claude Code / Codex: proposes likely next moves
  ├── Cactus Needle: proposes likely next moves
  └── Rules / active skills: add eligible workflow moves
        │
        ▼
Normalize all proposals into canonical trigger IDs
        │
        ▼
Candidate pool + metadata + provenance
        │
        ▼
Personalized ranker / policy selects 3 diverse choices
        │
        ▼
User selects one, rejects all, or enters custom action
        │
        ▼
Log choice, actual action, and outcome
  ├── Train Cactus Needle on human-validated actions
  └── Optionally distill high-quality frontier behavior
```

This makes Claude/Codex a source of strong priors and broad coverage, while the local model increasingly learns personalized preferences, project conventions, and operational habits.

---

### 1. Teacher-Generated Candidate Actions
Ask the frontier model to return a constrained structured response (e.g. 3–10 canonical action IDs, confidence, and concise evidence references):

```json
{
  "proposals": [
    {
      "trigger_id": "inspect_existing_pattern",
      "confidence": 0.79,
      "reason": "The repository likely already has an OAuth state-validation flow."
    },
    {
      "trigger_id": "write_failing_test",
      "confidence": 0.66,
      "reason": "The requirements are clear enough to lock in expected behavior."
    },
    {
      "trigger_id": "reproduce_and_trace_failure",
      "confidence": 0.44,
      "reason": "The failure may depend on callback/session state."
    }
  ]
}
```

Then combine frontier candidates with:
- Existing Skill triggers
- Candidates produced by Cactus Needle
- Deterministic candidates from workflow state (e.g. `run_targeted_tests` after editing code, `review_diff` before committing)
- Explicit "ask a clarifying question" and "none of these" actions.
- *Rule:* Map every proposal to a controlled action taxonomy first; do not allow free-form teacher prose to become a label automatically.

### 2. Synthetic-Label Distillation
When historical SDLC contexts exist with few human-labeled next actions, send those contexts to Claude/Codex to label the likely next move as a synthetic training set for Cactus Needle:

```text
Historical context → Frontier teacher predicts next-action distribution → Validate / filter output → Fine-tune Cactus Needle
```

Example structure:
```json
{
  "context": {
    "phase": "debugging",
    "signals": [
      "CI fails only on Linux",
      "error is a path-case mismatch",
      "local macOS test suite passes"
    ]
  },
  "teacher_target": {
    "trigger_id": "reproduce_in_target_environment",
    "alternatives": [
      "inspect_ci_logs",
      "add_regression_test"
    ],
    "confidence": 0.87
  },
  "label_source": "frontier_teacher"
}
```

*Caveat:* This is pseudo-labeling. Distillation methods can convert teacher-vs-student generations into preference pairs using a ranking objective rather than pure imitation.

### 3. Frontier Model as a Judge or Critic
Use the frontier model to evaluate a candidate pool generated by Cactus Needle, heuristics, and the Skill router:

```text
Candidate pool → Cactus Needle ranks candidates → Claude/Codex critiques/reranks candidates → UI shows 3 choices → User choice & outcome become final supervision
```

Critique fields: `feasibility`, `expected_information_gain`, `risk`, `reversibility`, `likely_unblocks_task`, `preconditions_missing`, `confidence`, `reason`.

---

## Signal Sources & Label Trust Weights

Do not merge all examples into an undifferentiated field. Track provenance and weight signals appropriately:

| Signal Source | Meaning | Typical Trust Weight |
|---|---|---|
| **Executed / Canonicalized user reply** | Operator actually took this move | **Highest (1.00)** |
| **Explicit user selection** | Operator preferred an offered option | **High (0.80)** |
| **Positive downstream outcome** | Action produced verified progress/success | **High, but delayed (0.60)** |
| **Frontier teacher proposal** | Claude/Codex considers it likely | **Medium (0.35)** |
| **Rule-generated candidate** | Deterministic workflow suggestion | **Low to medium (0.10)** |
| **Cactus Needle proposal** | Student prediction before feedback | **Not a training target by itself** |
| **User rejects all choices** | Candidate set missed the action | **Strong negative signal for shown slate** |

**Core Hierarchy:**
human action + outcome > human selection > teacher guess > heuristic

---

## Training Objectives

1. **Action Imitation (Behavioral Cloning):**
   Predict the normalized action actually taken:
   `L_BC = - sum_t w_t log pi_theta(a_executed_t | s_t)`
2. **Pairwise Preference Ranking:**
   If user selects one of three choices: `(s_t, a_chosen > a_rejected)`
   If user rejects all and executes a custom move: `(s_t, a_executed > a_1)`, `(s_t, a_executed > a_2)`, `(s_t, a_executed > a_3)`
3. **Teacher-Assisted Ranking:**
   Retain teacher candidate ordering `a_teacher_1 > a_teacher_2 > ...` as low/medium-weight synthetic preference data.
4. **Outcome-Weighted Reward:**
   `r_t = alpha * targeted_tests_pass + beta * task_advances - gamma * rework - delta * rollback`

---

## Product Pattern: Teacher Proposes, Human Arbitrates

1. **Capture state:** task, files, diffs, test status, recent tool calls, SDLC stage, repo conventions, active Skill.
2. **Ask Claude Code / Codex:** "Return top 5 next trigger IDs from allowed taxonomy with brief rationale and calibrated confidence."
3. **Ask Cactus Needle:** "Predict top 5 next trigger IDs and confidence from same state."
4. **Merge & Deduplicate:** Disallow invalid actions, preserve viable minority options, include exploration, ensure no destructive actions auto-run.
5. **Rank Pool:** Diverse scoring covering Information-gathering, Verification, and Intervention.
6. **Show 3 Diversified Choices** + "Other / type next move" + "Why these?" + explicit no-op/defer.
7. **Log:** Selected option, actual executed action, and subsequent outcome.

---

## Avoiding Selection & Logging-Policy Bias

To make offline contextual bandit evaluation and future policy learning sound:
- Log every candidate considered in the candidate pool, not merely the displayed three.
- Log display position, propensity, and scores.
- Log proposing model/source, prompt template version, and model version.
- Inject controlled exploration (e.g. reserving 1 slot for valid lower-ranked candidates).
- Preserve user typed overrides as positive behavioral signal.

---

## Suggested Event Schema

```json
{
  "decision_id": "uuid",
  "timestamp": "2026-09-13T16:49:00-07:00",
  "context": {
    "task_id": "local-task-or-hash",
    "sdlc_phase": "debugging",
    "repo_state": {
      "changed_files": ["src/auth/callback.ts"],
      "tests": {
        "targeted": "failing",
        "failure_class": "invalid OAuth state accepted"
      }
    },
    "recent_actions": [
      "inspect_failure",
      "open_callback_handler"
    ]
  },
  "action_taxonomy_version": "sdlc-triggers-v1",
  "proposals": [
    {
      "trigger_id": "inspect_existing_pattern",
      "source": "claude_code",
      "model": "claude-code-model-version",
      "rank": 1,
      "confidence": 0.81,
      "prompt_version": "next-move-v3"
    },
    {
      "trigger_id": "write_failing_test",
      "source": "codex",
      "model": "codex-model-version",
      "rank": 1,
      "confidence": 0.76,
      "prompt_version": "next-move-v3"
    },
    {
      "trigger_id": "implement_minimal_fix",
      "source": "cactus_needle",
      "model": "cactus-needle-28b-r1",
      "rank": 2,
      "confidence": 0.63
    }
  ],
  "candidate_pool": [
    "inspect_existing_pattern",
    "write_failing_test",
    "implement_minimal_fix",
    "reproduce_in_target_environment"
  ],
  "displayed_choices": [
    {
      "trigger_id": "inspect_existing_pattern",
      "display_position": 1,
      "selection_propensity": 0.62,
      "sources": ["claude_code", "workflow_rule"]
    },
    {
      "trigger_id": "write_failing_test",
      "display_position": 2,
      "selection_propensity": 0.25,
      "sources": ["codex", "cactus_needle"]
    },
    {
      "trigger_id": "reproduce_in_target_environment",
      "display_position": 3,
      "selection_propensity": 0.13,
      "sources": ["exploration"]
    }
  ],
  "human_feedback": {
    "selected_trigger_id": "inspect_existing_pattern",
    "raw_reply": "Check how the existing refresh flow validates state.",
    "actual_trigger_id": "inspect_existing_pattern",
    "selection_mode": "explicit_choice",
    "canonicalization_confidence": 0.98
  },
  "outcome": {
    "completed": true,
    "targeted_tests_passed": true,
    "reverted": false,
    "rework_count": 0,
    "outcome_observed_at": "2026-09-13T17:09:00-07:00"
  }
}
```

---

## Model Role Taxonomy

| Component | Canonical Role Name |
|---|---|
| Claude Code / Codex | Frontier teacher, proposer, or critic |
| Cactus Needle 28B | Local student, personalized policy, or re-ranker |
| Claude Skills / 10 Workflow Actions | Action taxonomy, skill catalog, or action space |
| Operator Choice & Actual Action | Human demonstration / ground-truth behavioral label |
| Tests, CI, PR status, rollback, rework | Outcome and reward signals |
| UI 3-Option Selector | Candidate selector / slate policy |

---

## Practical Staged Sequencing

1. **Structured Proposals:** Have Claude/Codex generate 5 canonical trigger candidates with structured confidence and rationale.
2. **Student Proposals:** Have Cactus Needle independently generate top 5.
3. **Diversified Slate:** Merge candidates, displaying three intentionally diverse, valid options (Info-gathering, Verification, Intervention).
4. **Comprehensive Logging:** Log explicit selection, typed override, executed tool/skill, and downstream outcome.
5. **Local Preference Ranker:** Train an inexpensive local ranker first on human preference pairs.
6. **LoRA Fine-Tuning:** Add LoRA/QLoRA fine-tuning of Cactus Needle only after accumulating a solid base of reviewed, canonicalized human-action examples.
7. **Disjoint Holdouts:** Use teacher examples strictly for bootstrap data; evaluate on a fixed, human-labeled holdout split by repository/task/time.

**Comment by noelsaw1** [comment]
Synthesized next-step research and integration plans with our audited SDLC skills and ModernBERT benchmarks are tracked in canonical issue [#547 (comment)](https://github.com/HiQS-Labs/XYZ-forge/issues/547#issuecomment-5657713339).
